iroh-rpc/src/core.rs

```diff
@@ -0,0 +1,585 @@
+                            .send(SenderType::Stream {
+                                header,
+                                stream: receiver,
+                            });
+                    }
+                    StreamingResponseEvent::Ack => {
+                        debug!(target: "inbound streaming", "Received Ack for request {}", request_id);
+                        let _ = self
+                            .pending_requests
+                            .remove(&PendingId::RequestId(request_id))
+                            .expect("Request to still be pending.")
+                            .send(SenderType::Ack);
+                    }
+                    StreamingResponseEvent::StreamError(e) => {
+                        debug!(target: "inbound streaming", "Received Ack for request {}", request_id);
+                        let _ = self
+                            .pending_requests
+                            .remove(&PendingId::RequestId(request_id))
+                            .expect("Request to still be pending.")
+                            .send(SenderType::Error(e.into()));
+                    }
+                },
+            },
+            StreamingEvent::OutboundFailure {
+                request_id, error, ..
+            } => {
+                let _ = self
+                    .pending_requests
+                    .remove(&PendingId::RequestId(request_id))
+                    .expect("Request to still be pending.")
+                    .send(SenderType::Error(Box::new(error)));
+            }
+            StreamingEvent::InboundFailure {
+                peer,
+                request_id,
+                error,
+            } => {
+                print!("\n\nThis peer_id: {}\n", self.swarm.local_peer_id());
+                println!(
+                    "RequestResponseEvent::InboundFailure:\npeer_id: {}\nrequest_id: {}\nerror: {}",
+                    peer, request_id, error
+                );
+            }
+            StreamingEvent::ResponseSent { .. } => {}
+        }
+    }
+}
+
+/// An InboundEvent is translated from an event that comes over the network
+/// that the process needs to handle.
+pub struct InboundEvent {
+    pub command: InCommand,
+    pub channel: StreamingResponseChannel,
+}
+
+/// Build a swarm with an in memory transport.
+// TODO: eventually generalize, pass in config etc
+pub fn new_mem_swarm(id_keys: Keypair) -> Swarm<CoreBehaviour> {
+    let peer_id = id_keys.public().to_peer_id();
+    SwarmBuilder::new(mem_transport(id_keys), CoreBehaviour::new(), peer_id).build()
+}
+
+/// Build a swarm with a TCP transport
+// TODO: eventually generalize, pass in config etc
+pub async fn new_swarm(id_keys: Keypair) -> Result<Swarm<CoreBehaviour>, Box<dyn Error>> {
+    let peer_id = id_keys.public().to_peer_id();
+    Ok(SwarmBuilder::new(
+        libp2p::development_transport(id_keys).await?,
+        CoreBehaviour::new(),
+        peer_id,
+    )
```

Comment:
this is missing setting up tokio as the executor, this is pretty important: https://github.com/libp2p/rust-libp2p/blob/master/examples/chat-tokio.rs#L140-L144